kotlinx-coroutines-core/jvm/src/CoroutineExceptionHandlerImpl.kt

```diff
@@ -13,10 +13,15 @@ import kotlin.coroutines.*
  * Note that Android may have dummy [Thread.contextClassLoader] which is used by one-argument [ServiceLoader.load] function,
  * see (https://stackoverflow.com/questions/13407006/android-class-loader-may-fail-for-processes-that-host-multiple-applications).
  * So here we explicitly use two-argument `load` with a class-loader of [CoroutineExceptionHandler] class.
+ *
+ * We are explicitly using the `ServiceLoader.load(MyClass::class.java, MyClass::class.java.classLoader).iterator()`
+ * form of the ServiceLoader call to enable R8 optimization when compiled on Android.
  */
-private val handlers: List<CoroutineExceptionHandler> = CoroutineExceptionHandler::class.java.let { serviceClass ->
-    ServiceLoader.load(serviceClass, serviceClass.classLoader).toList()
-}
+private val handlers: List<CoroutineExceptionHandler> = ServiceLoader.load(
+        CoroutineExceptionHandler::class.java,
+        CoroutineExceptionHandler::class.java.classLoader
+).iterator().asSequence().toList()
```

Comment:
Just out of curiosity, could you please explain why `ServiceLoader.load(...).iterator()...` enables optimization, while `ServiceLoader.load(...).anythingElse()` does not?

Reply:
I wasn't involved in the design of the R8 optimization, but looking at the source code of R8, these are the necessary conditions. My guess is because they're actually replacing the ServiceLoader entirely with direct instantiation and putting the loaded services into a Collection (if I remember correctly), they want to make sure that only code which accesses the Iterator (which both the Collection and ServiceLoader implement) is safe to modify in this way. 

If an app did something else like `ServiceLoader.getClass()` (for whatever reason), then there would be a mismatch between the optimized and non-optimized version.

Reply:
Yeah it was meant to handle this Java pattern:
```java
for (Service service : ServiceLoader.loader(Service.class, Service.class.getClassLoader()) {
  // ...
}
```
Unfortunately Kotlin's nice extensions like `toList()` have the `ServiceLoader` passed as an instance to a function rather than inlining `iterator()` to the call site. `.forEach` and friends would still work though.

We _might_ be able to relax the optimization where it sees that the `ServiceLoader` flows into a function as an `Iterable`. Although, as @wojtek-kalicinski states, someone could still be doing `getClass()` or `toString()` or, for some horrific reason, `instanceof ServiceLoader`.